### PR TITLE
Updating Module Tracker

### PR DESCRIPTION
**Pull request by sanketpurandare**
Add functionality to register and clear user defined hooks to Module Tracker.
Add utility method to get module fqn (fully qualified name) stored by Module Tracker.

@albanD

### CODE REVIEW

File: torch/utils/module_tracker.py [hunk 1]
```diff
@@ -76,6 +77,74 @@ def is_bw(self):
...
+    def register_user_hooks(
+        self,
+        pre_fw_hook: Optional[Callable] = None,
+        post_fw_hook: Optional[Callable] = None,
+        pre_bw_hook: Optional[Callable] = None,
+        post_bw_hook: Optional[Callable] = None,
+    ):
+        """
+        Registers user-specified hooks to be called before/after the forward/backward pass for each
+        module tracked by the Module Tracker. One or more can be none.
```
Comment (albanD) [inline torch/utils/module_tracker.py:95]:
```suggestion
        module tracked by the Module Tracker. One or more can be ``None``.
```

File: torch/utils/module_tracker.py [hunk 2]
```diff
@@ -76,6 +77,74 @@ def is_bw(self):
...
+
+    def register_user_hooks(
+        self,
+        pre_fw_hook: Optional[Callable] = None,
+        post_fw_hook: Optional[Callable] = None,
+        pre_bw_hook: Optional[Callable] = None,
+        post_bw_hook: Optional[Callable] = None,
+    ):
+        """
+        Registers user-specified hooks to be called before/after the forward/backward pass for each
```
Comment (albanD) [inline torch/utils/module_tracker.py:94]: We should mention here that when the hooks called, the current module is always inside the parents set

**Review by albanD — COMMENTED** [review]
Sounds good.
Can you add some tests to make sure this all works fine?

File: torch/utils/module_tracker.py (hunk 1 above)
Comment (sanketpurandare) [reply torch/utils/module_tracker.py:95]: done

File: torch/utils/module_tracker.py (hunk 2 above)
Comment (sanketpurandare) [reply torch/utils/module_tracker.py:94]: I have already mentioned that in 

> Note:
            If the module is not alive during the backward pass, the pre_bw_hook and post_bw_hook will
            will receive None as the module argument.
            The module fqn will be present in the ``parents`` attribute when each of the hooks is called.
            Hooks are intended to be used as markers only not to modify the inputs/outputs.

Comment (albanD) [reply torch/utils/module_tracker.py:94]: Ho my bad, sounds good!

File: test/test_module_tracker.py [hunk 3]
```diff
@@ -76,6 +76,64 @@ def test_bw_detection(self):
             self.assertFalse(tracker.is_bw)
             self.assertEqual(tracker.parents, {"Global"})
 
+    def test_user_hooks(self):
+        class Bar(torch.nn.Module):
+            def __init__(self):
+                super(__class__, self).__init__()
```
Comment (albanD) [inline test/test_module_tracker.py:82]:
```suggestion
                super().__init__()
```
Not needed anymore in 3+

File: test/test_module_tracker.py [hunk 4]
```diff
@@ -76,6 +76,64 @@ def test_bw_detection(self):
...
+                self.foo = torch.nn.Linear(10, 10)
+
+            def forward(self, x):
+                return self.foo(x).relu_()
+
+        mt = ModuleTracker()
+        test_op = []
+
+        def hook(mod, mt, hook_name):
```
Comment (albanD) [inline test/test_module_tracker.py:91]: Given how this looks like, should the first argument passed to the hook always be the module tracker?

Comment (sanketpurandare) [reply test/test_module_tracker.py:91]: This is just for testing? The actual hooks have a clear signature:

1. pre_fw(mod, inp)
2. post_fw(mod, inp, op)
3. pre_bw(mod, grad_op)
4. post_bw(mod, grad_inp)

Comment (sanketpurandare) [reply test/test_module_tracker.py]: Actually, I just modified the hook to remove any confusion.

**Pull request closed without merging**